### DISCUSSION TITLE
No PR for Submodules

### DISCUSSION THREAD

**Discussion by ekristen** [renovatebot/renovate#9367] [Request Help]
I found another question/answer on here but it doesn't seem to actually answer the problem with a solution exactly.

I have two modules both within the same org, both integrated with the renovate app and both private.

One submodules the other, there have been new updates to the submoduled repo, but not PR has been opened.

We do have `.github/renovate.json` set with git-submodules enabled.

Thoughts?

**Comment by rarkins** [comment]
Start with the logs, do you see anything to indicate partial success or failure? Eg search for "modules" string within the logs and then explore the hits

**Reply by ekristen** [reply]
Didn't even realize until now that they were accessible! Checking now.

**Reply by ekristen** [reply]
```Repository not found.\nfatal: Could not read from remote repository.```

**Reply by ekristen** [reply]
There was mention on the other discussion about needing to modify the git config, but it seems like it was mixed on whether that was necessary or not. 

There was also mention on hosts rules, but it wasn't clear if that was needed or not either.

**Comment by ekristen** [comment]
Logs indicate permission issue with the submodule private repository. So I guess the question is when it comes to the renovate app, what needs to be done to ensure private submodules work when they are A) within the same org, B) in separate orgs. 

Is this where host rules come in? Do we then have to use the oauth url rewrite trick for github urls?

**Reply by rarkins** [reply]
I forget exactly if/how private repo lookups work for submodules. Could you start by finding the entry for it in the "packageFiles with updates" log message? e.g. to identify datasource, depName, etc.

**Reply by ekristen** [reply]
Interesting it detects 
```
{
  "baseBranch": "master",
  "stats": {
    "managers": {
      "git-submodules": {
        "fileCount": 1,
        "depCount": 0
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 0
    }
  }
}
```

but then has zero deps under packageFiles

```
packageFiles with updates
{
  "config": {
    "git-submodules": [
      {
        "packageFile": ".gitmodules",
        "deps": [],
        "datasource": "git-refs"
      }
    ]
  }
}
```

During clone is when it hits the unable to initialize/clone submodule error.

**Reply by ekristen** [reply]
@rarkins when you get a moment see the above, appreciate the assist!

**Reply by rarkins** [reply]
So it indicates that it found the *file* but did not identify any module definition within. Can you paste the contents here, sanitizing first if necessary?

**Reply by ekristen** [reply]
```
INFO: Repository started
{
  "renovateVersion": "24.103.0"
}
DEBUG: Using localDir: /mnt/renovate/gh/org/test-submodule
DEBUG: Repository cache is valid
DEBUG: initRepo("org/test-submodule")
DEBUG: Overriding default GitHub endpoint
{
  "endpoint": "https://api.github.com/"
}
DEBUG: No dangling containers to remove
DEBUG: org/test-submodule default branch = master
DEBUG: Using app token for git init
DEBUG: resetMemCache()
DEBUG: checkOnboarding()
DEBUG: isOnboarded()
DEBUG: findFile(renovate.json)
DEBUG: Initializing git repository into /mnt/renovate/gh/org/test-submodule
DEBUG: git clone completed
{
  "durationMs": 851
}
DEBUG: Cloning git submodule at deps/demo
WARN: Unable to initialise git submodule at deps/demo
{
  "err": {
    "task": {
      "commands": [
        "submodule",
        "update",
        "--init",
        "deps/demo"
      ],
      "format": "utf-8"
    },
    "message": "Submodule 'deps/demo' (git@github.com:org/test-submodule-base.git) registered for path 'deps/demo'\nCloning into '/mnt/renovate/gh/org/test-submodule/deps/demo'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\nfatal: clone of 'git@github.com:org/test-submodule-base.git' into submodule path '/mnt/renovate/gh/org/test-submodule/deps/demo' failed\nFailed to clone 'deps/demo'. Retry scheduled\nCloning into '/mnt/renovate/gh/org/test-submodule/deps/demo'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\nfatal: clone of 'git@github.com:org/test-submodule-base.git' into submodule path '/mnt/renovate/gh/org/test-submodule/deps/demo' failed\nFailed to clone 'deps/demo' a second time, aborting\n",
    "stack": "Error: Submodule 'deps/demo' (git@github.com:org/test-submodule-base.git) registered for path 'deps/demo'\nCloning into '/mnt/renovate/gh/org/test-submodule/deps/demo'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\nfatal: clone of 'git@github.com:org/test-submodule-base.git' into submodule path '/mnt/renovate/gh/org/test-submodule/deps/demo' failed\nFailed to clone 'deps/demo'. Retry scheduled\nCloning into '/mnt/renovate/gh/org/test-submodule/deps/demo'...\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\nfatal: clone of 'git@github.com:org/test-submodule-base.git' into submodule path '/mnt/renovate/gh/org/test-submodule/deps/demo' failed\nFailed to clone 'deps/demo' a second time, aborting\n\n    at Object.action (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/plugins/error-detection.plugin.js:30:33)\n    at PluginStore.exec (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/plugins/plugin-store.js:21:33)\n    at /home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:92:45\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:90:16)\n    at GitExecutorChain.<anonymous> (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:74:46)\n    at Generator.next (<anonymous>)\n    at fulfilled (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)"
  }
}
DEBUG: latest repository commit
{
  "latestCommit": {
    "hash": "080acc9381089dc4e27f5386409d8b9e63e5c7d1",
    "date": "2021-04-01T14:14:34-06:00",
    "message": "adding renovate for submodules",
    "refs": "HEAD -> master, origin/master, origin/HEAD",
    "body": ""
  }
}
DEBUG: Setting git author name
{
  "gitAuthorName": "Renovate Bot"
}
DEBUG: Setting git author email
{
  "gitAuthorEmail": "bot@renovateapp.com"
}
DEBUG: findFile(renovate.json5)
DEBUG: findFile(.github/renovate.json)
DEBUG: Config file exists
{
  "fileName": ".github/renovate.json"
}
DEBUG: Retrieving issueList
DEBUG: Retrieved 0 issues
DEBUG: Repo is onboarded
DEBUG: Found .github/renovate.json config file
DEBUG: Repository config
{
  "fileName": ".github/renovate.json",
  "config": {
    "git-submodules": {
      "enabled": true
    }
  }
}
DEBUG: migrateAndValidate()
DEBUG: No config migration necessary
DEBUG: massaged config
{
  "config": {
    "git-submodules": {
      "enabled": true
    },
    "extends": [
      "github>whitesource/merge-confidence:beta"
    ]
  }
}
DEBUG: migrated config
{
  "config": {
    "git-submodules": {
      "enabled": true
    },
    "extends": [
      "github>whitesource/merge-confidence:beta"
    ]
  }
}
DEBUG: Found repo ignorePaths
{
  "ignorePaths": [
    "**/node_modules/**",
    "**/bower_components/**"
  ]
}
DEBUG: detectSemanticCommits()
DEBUG: getCommitMessages
DEBUG: Semantic commits detection: unknown
DEBUG: No semantic commits detected
DEBUG: Setting branchPrefix: renovate/
DEBUG: Cannot read vulnerability alerts
DEBUG: No vulnerability alerts found
DEBUG: processRepo()
DEBUG: No baseBranches
DEBUG: extract()
DEBUG: Found cached extract
{
  "baseBranch": "master",
  "baseBranchSha": "080acc9381089dc4e27f5386409d8b9e63e5c7d1"
}
DEBUG: Deleted cached dep updates
INFO: Dependency extraction complete
{
  "baseBranch": "master",
  "stats": {
    "managers": {
      "git-submodules": {
        "fileCount": 1,
        "depCount": 0
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 0
    }
  }
}
DEBUG: Package releases lookups complete
{
  "baseBranch": "master"
}
DEBUG: branchifyUpgrades
DEBUG: 0 flattened updates found:
DEBUG: Returning 0 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "config": {
    "git-submodules": [
      {
        "packageFile": ".gitmodules",
        "deps": [],
        "datasource": "git-refs"
      }
    ]
  }
}
DEBUG: processRepo()
DEBUG: Processing 0 branches:
DEBUG: Calculated maximum PRs remaining this run
{
  "prsRemaining": 99
}
DEBUG: PullRequests limit = 99
DEBUG: Calculated maximum branches remaining this run
{
  "branchesRemaining": 99
}
DEBUG: Branches limit = 99
DEBUG: Removing any stale branches
DEBUG: config.repoIsOnboarded=true
DEBUG: No renovate branches found
DEBUG: Repository timing splits (milliseconds)
{
  "splits": {
    "init": 5454,
    "extract": 3,
    "lookup": 8,
    "update": 0
  },
  "total": 5500
}
DEBUG: http statistics
{
  "hostStats": [
    "api.github.com, 4 requests, 208ms request average, 0ms queue average"
  ],
  "totalRequests": 4
}
INFO: Repository finished
{
  "durationMs": 5500
}
```

**Reply by rarkins** [reply]
Curiously it hasn't reported any sub modules deps array even though it attempts to find them.

**Reply by ekristen** [reply]
I think it's because it fails to clone it to begin with? I'm guessing I need to do something to allow the Renovate GH Action to clone the private submodule?

**Reply by rarkins** [reply]
Here's the code: https://github.com/renovatebot/renovate/blob/486f5747f60d05cad90989039443d94f43aaf8ef/lib/manager/git-submodules/extract.ts#L61-L88

**Reply by ekristen** [reply]
That command seems to be valid and it should be finding the submodules regardless of the clone error..

**Comment by ekristen** [comment]
@rarkins it seems like it's simply a permissions issue, is there documentation on the best/approved/correct way to ensure the renovate bot can access the other private repositories especially when they are referenced via ssh? Thanks!

**Reply by rarkins** [reply]
Impossible over ssh if you use the hosted app. If you self host, you'd want to make sure the current user has an ssh profile which can access what you need.

**Reply by ekristen** [reply]
I've seen places where people suggest rewriting the urls to use a token? Is that possible? I suppose I can try and figure out how to run renovate as a GitHub action as an alternative.

**Reply by rarkins** [reply]
I think it's possible to use a syntax for the remote repo so that it's possible to use ssh locally but for renovate to use https instead. @JamieMagee do you recall?

**Reply by ekristen** [reply]
Would like to follow up, @JamieMagee is this possible? Thank you. If not I'll just look into switching to self-hosted or github-action hosted instead.

**Reply by JamieMagee** [reply]
Yes, it should be possible to keep the `ssh` reference in `.gitmodules`, and renovate should rewrite that to HTTPS. However, you'll need to configure a hostrule with a token that Renovate can use to clone the submodule. [Here's our documentation for `hostRules`](https://docs.renovatebot.com/configuration-options/#hostrules)

**Reply by ekristen** [reply]
@JamieMagee do you happen to have an example somewhere you can link to?

**Reply by ekristen** [reply]
I just found someone said this should work https://github.com/renovatebot/renovate/discussions/8529#discussioncomment-604705

Follow up question, can the value of token be an environment variable?

**Reply by ekristen** [reply]
I found this @JamieMagee but no hostRules defined? https://github.com/JamieMagee/renovate-git-submodule-repro

**Reply by ekristen** [reply]
Trying host rules 

```
{
  "git-submodules": {
    "enabled": true
  },
  "hostRules": [
    {
      "matchHost": "github.com",
      "password": "x-oauth-basic",
      "encrypted": {
        "username": "<encrypted>"
      }
    }
  ]
}
```

```
Unable to initialise git submodule at <repo>

Submodule '<repo>' (git@github.com:<repo>.git) registered for path '<repo>'\nCloning into '/mnt/renovate/gh/sans-sroc/<repo_parent>/<repo>'...\nERROR: Repository not found
```

**Reply by ekristen** [reply]
I've also tried encryped.token as well.

**Comment by ekristen** [comment]
@JamieMagee not sure if thread comments will get someone's attention, been 2 days, so pinging you here. Please see my above comments.

TL;DR -- I've seen your source repo for testing but see no host rules, assume you tested with self-hosted? Can this be done with the cloud hosted version or am I out of luck?

**Reply by JamieMagee** [reply]
Hey, sorry for the lack of response. I've been busy with other things.

All of my testing and running in production has been with self-hosted renovate. I've not tested with the cloud hosted version. It should be possible, but may require code changes.

From the logs you shared above it looks like we're not correctly rewriting an SSH URL (`git@github.com:<repo>.git`) into an HTTPS url (`https://github.com/<repo>.git`). First step for me will be to get a reproduction.
